### PR TITLE
RFR- add review controller and views

### PR DESCRIPTION
**Pull request by sammcgrail**

### CODE REVIEW

File: app/controllers/reviews_controller.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
+class ReviewsController < ApplicationController
+  def new
+    @rubygem = Rubygem.find(params[:rubygem_id])
+    @review = Review.new
+  end
+
+  def create
+    @rubygem = Rubygem.find(params[:rubygem_id])
+    @review = Review.new(review_params)
+    @review.rubygem = @rubygem
```
Comment (davidrf) [inline app/controllers/reviews_controller.rb]: Use associations to create your new review object

```ruby
  @review = @rubygem.reviews.new(review_params)
```

File: app/controllers/reviews_controller.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    @review.user = current_user
+
+    if @review.save
+      flash[:notice] = 'Review added.'
+      redirect_to rubygem_path(@rubygem.id)
```
Comment (davidrf) [inline app/controllers/reviews_controller.rb]: You can just give path helpers objects

```ruby
rubygem_path(@rubygem)
```

File: spec/features/user_adds_a_review_spec.rb [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    visit new_user_session_path
+    sign_in_as(user)
+    visit rubygem_path(rubygem)
+
+    click_link 'Add Review'
+
+    select('4', from: "Rating")
+    fill_in 'Body', with: 'test review body'
+    click_button 'Submit'
+    expect(page).to have_content('test review body')
```
Comment (davidrf) [inline spec/features/user_adds_a_review_spec.rb]: Minor reformatting. You should have a line break between your setup (FactoryGirl declarations if needed), the user interface, and your expectations. Example:

```ruby
+    user = FactoryGirl.create(:user)
+    rubygem = FactoryGirl.create(:rubygem)
+
+    visit new_user_session_path
+    sign_in_as(user)
+    visit rubygem_path(rubygem)
+    click_link 'Add Review'
+    select('4', from: "Rating")
+    fill_in 'Body', with: 'test review body'
+    click_button 'Submit'
+
+    expect(page).to have_content('test review body')
```

File: spec/features/user_adds_a_review_spec.rb [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  5 star rating, with an optional body
+  So that I can share feedback about a gem
+} do
+  # Acceptance Criteria
+  # [  ] When I click "review gem", I can submit a star rating
+  # [  ] When I click "review gem", I can optionally submit a review
+
+  scenario 'authenticated user adds a review' do
+    user = FactoryGirl.create(:user)
+    rubygem = FactoryGirl.create(:rubygem)
```
Comment (davidrf) [inline spec/features/user_adds_a_review_spec.rb]: You're using rubygem in multiple tests, you should use a let statement that is outside both scenarios but still within the feature block.

File: spec/features/user_adds_a_review_spec.rb [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    click_button 'Submit'
+    expect(page).to have_content('test review body')
+  end
+
+  scenario 'unauthenticated user attempts to add a review' do
+    rubygem = FactoryGirl.create(:rubygem)
+    visit rubygem_path(rubygem)
+
+    click_link 'Add Review'
+    expect(page).to have_content('You must sign up in order to leave a review!')
```
Comment (davidrf) [inline spec/features/user_adds_a_review_spec.rb]: Format as above

File: app/controllers/reviews_controller.rb [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    else
+      flash[:notice] = 'Review was not added.'
+      render :new
+    end
+  end
+
+  def show
+    @review = Review.find(params[:id])
+    ## FIX
+  end
```
Comment (davidrf) [inline app/controllers/reviews_controller.rb]: You need another test for review show page in a new file. Basically this test will start out at the index page, then click the appropriate link which will take you to the show page and you should expect to see the specific details of the ruby gem.

File: spec/features/view_all_gems_spec.rb [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+  # [ ] I can see all gems on the home page
+
+  scenario 'unauthorized user views all gems on home page' do
+    rubygem = FactoryGirl.create(:rubygem)
+    review = FactoryGirl.create(:review, rubygem_id: rubygem.id)
+
+    visit rubygems_path
+    expect(page).to have_content("ALL GEMS")
+    expect(page).to have_content(rubygem.name)
+
```
Comment (davidrf) [inline spec/features/view_all_gems_spec.rb]: This is the index feature of rubygems, your test should end here (Also, you would get rid of the reviews).

File: spec/features/view_all_gems_spec.rb [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    click_link(rubygem.name)
+    expect(page).to have_content(review.rating)
+    expect(page).to have_content(review.body)
+    expect(page).to have_content(review.user.username)
```
Comment (davidrf) [inline spec/features/view_all_gems_spec.rb]: Create another file (call it reviews_index) that includes these expectations lines (23-25). Before these expectations a user would go to the rubygems index, click the name of the of the review gem. Here you would need to create a review object.

File: spec/features/view_all_gems_spec.rb [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+  scenario 'unauthorized user views all gems on home page' do
+    rubygem = FactoryGirl.create(:rubygem)
+    review = FactoryGirl.create(:review, rubygem_id: rubygem.id)
+
+    visit rubygems_path
+    expect(page).to have_content("ALL GEMS")
+    expect(page).to have_content(rubygem.name)
+
+    click_link(rubygem.name)
```
Comment (davidrf) [inline spec/features/view_all_gems_spec.rb]: You are still missing another file (call it rubygems_show_spec.rb) where a user views the index page, clicks on the name of the ruby gem, and then sees all the details of a ruby gem. You do not have to check for existence of reviews.

File: app/models/rubygem.rb [hunk 10]
```diff
@@ -5,4 +5,19 @@ class Rubygem < ActiveRecord::Base
   validates :user_id, presence: true
   validates :link_to_rubygems_url, presence: true
   validates :link_to_rubygems_url, format: URI::regexp(%w(http https))
+
+  def average_rating
+    total = []
```
Comment (vikram7) [inline app/models/rubygem.rb]: can't total be an integer? `total = 0` and then `total += review.rating` ?

File: app/models/rubygem.rb [hunk 11]
```diff
@@ -5,4 +5,19 @@ class Rubygem < ActiveRecord::Base
...
+    Rubygem.all.each do |rubygem|
```
Comment (vikram7) [inline app/models/rubygem.rb]: I'm a bit confused. Is this an average rating for a single ruby gem or for all rubygems?

File: app/views/reviews/show.html.erb [hunk 12]
```diff
@@ -0,0 +1,8 @@
+<div class="row">
+  <div class="large-12">
+    <h1><%= @review.user %></h3>
+    <%= @review.rating %>
+    <%= @review.body %>
```
Comment (vikram7) [inline app/views/reviews/show.html.erb]: lines 4 and 6 need `<div>` tags around them.

File: app/views/rubygems/show.html.erb [hunk 13]
```diff
@@ -1,9 +1,17 @@
...
-    <%= @rubygem.link_to_rubygems_url %>
-    <br>
+    <h1><%= @rubygem.name %></h1>
+    <h3>URL: <%= @rubygem.link_to_rubygems_url %></h3>
+    <h3>Average rating: <%= @rubygem.average_rating %></h3>
     <br>
     <h3>Reviews</h3>
+    <% @reviews.each do |review| %>
+    <ul>
+      <li>Rating: <%= review.rating %></li><br>
```
Comment (vikram7) [inline app/views/rubygems/show.html.erb]: You should have `<div>` tags instead of `<br>` tags here.

File: app/views/rubygems/index.html.erb [hunk 14]
```diff
@@ -1,4 +1,22 @@
+<<<<<<< HEAD
```
Comment (davidrf) [inline app/views/rubygems/index.html.erb]: Get rid of this

Comment (davidrf) [inline app/views/rubygems/index.html.erb]: Dealt with this

File: app/views/rubygems/index.html.erb [hunk 15]
```diff
@@ -1,4 +1,22 @@
...
+    <div class="six columns">
+      <%= text_field_tag(:search) %>
+    </div>
+    <div class="two columns">
+      <%= submit_tag("Search for Gems", class: "button expand postfix") %>
+    </div>
+  <% end %>
+</div>
+
+=======
```
Comment (davidrf) [inline app/views/rubygems/index.html.erb]: and this

File: app/views/rubygems/index.html.erb [hunk 16]
```diff
@@ -1,4 +1,22 @@
...
 <h1>List of all Gems</h1>
+>>>>>>> master
```
Comment (davidrf) [inline app/views/rubygems/index.html.erb]: Also this

File: spec/features/rubygems_index_spec.rb [hunk 17]
```diff
@@ -0,0 +1,19 @@
+require 'rails_helper'
+require 'pry-rails'
```
Comment (davidrf) [inline spec/features/rubygems_index_spec.rb]: You shouldn't need to require 'pry-rails' here

File: spec/features/rubygems_show_spec.rb [hunk 18]
```diff
@@ -0,0 +1,27 @@
+require 'rails_helper'
+require 'pry-rails'
```
Comment (davidrf) [inline spec/features/rubygems_show_spec.rb]: Same comment as above

File: app/controllers/reviews_controller.rb [hunk 19]
```diff
@@ -0,0 +1,33 @@
+class ReviewsController < ApplicationController
+  def new
+    @rubygem = Rubygem.find(params[:rubygem_id])
+    @review = Review.new
+  end
+
+  def create
+    @rubygem = Rubygem.find(params[:rubygem_id])
+    @review = @rubygem.reviews.new(review_params)
+    @review.rubygem = @rubygem
```
Comment (davidrf) [inline app/controllers/reviews_controller.rb]: This line is no longer be needed. Line 9 takes care of setting that association for you.

File: app/helpers/rubygems_helper.rb [hunk 20]
```diff
@@ -1,4 +1,15 @@
...
+    total = 0.00
+    rubygem.reviews.each do |review|
+      total += review.rating
+    end
+    if rubygem.reviews.all.count == 0
+      return 0
+    else
+      total / rubygem.reviews.all.count
+    end
+  end
```
Comment (davidrf) [inline app/helpers/rubygems_helper.rb]: This should be an instance method of a ruby gem. Also clean up your logic. You should return 0 right away if there are no reviews. If there are reviews then go ahead and calculate. Once you have all the reviews for that gem, you can just call `#average` method on the activerecord relation object.

Comment (davidrf) [inline app/helpers/rubygems_helper.rb]: Take this out

File: app/models/rubygem.rb [hunk 21]
```diff
@@ -29,4 +29,16 @@ def self.filter(params)
     end
     ruby_gems
   end
+
+  def average_rating(rubygem)
+    total = 0.00
+    if rubygem.reviews.all.count == 0
```
Comment (davidrf) [inline app/models/rubygem.rb]: No need for `#all` here

File: app/models/rubygem.rb [hunk 22]
```diff
@@ -29,4 +29,16 @@ def self.filter(params)
...
+      return 0
+    else
+      rubygem.reviews.each do |review|
```
Comment (davidrf) [inline app/models/rubygem.rb]: Use Rail's `#average` on all the reviews to calculate average rating.

**Pull request merged**